posts/benchmarking_design_decisions/index.qmd

```diff
@@ -0,0 +1,228 @@
+warnings <- c(1, 5, 10, 15, 20, 100)
+```
+
+Let's run benchmarks over the different scenarios and store results
+in a data.frame.
+
+```{r warning=FALSE}
+bnmark_list <- warnings |>
+  map(
+    \(x) mark(
+      "cli::cli_warn()" = warn_Ntimes(x, pkg = "cli"),
+      "base::warning()" = warn_Ntimes(x, pkg = "base")
+    )
+  )
+
+bnmark_tbl <- map2(
+  bnmark_list,
+  warnings,
+  \(x, y) mutate(x, warnings = y)
+) |>
+  reduce(bind_rows)
+```
+
+## Results
+
+It's time to explore the results of the data generated. We'll start by
+making a boxplot of the median run times for the different scenarios.
+
+Drum roll please...
+
+```{r}
+autoplot(bnmark_tbl, type = "boxplot") +
+  labs(
+    x = "Package",
+    caption = "Functions throwing various numbers of warnings as indicated in the facet label"
+  )
+```
+
+Finally, let's investigate the relative speeds of the two packages with
+respect to the number of warnings. We will do this by dividing the speed of
+`cli::cli_warn()` by the speed of `base::warning()`.
+
+```{r}
+bnmark_tbl |>
+  select(expression, median, warnings) |>
+  pivot_wider(
+    names_from = expression,
+    values_from = c(median)
+    ) |> 
+  mutate(
+    scale = `cli::cli_warn()`/`base::warning()`
+    )
+```
+
+As we can see, `cli` is consistently slower than base R but this seems not to
+scale with the number of warnings thrown. Other benchmarking experiments
+with `cli` have revealed linear relationships between its speed and the number
+of times it invokes other functions [^3]. The developers of `cli` have also
+conducted benchmarks of the `ansi_` family of functions in `cli` in comparison
+to base R and the `fansi` package. They find that `cli` is consistently slower
+than base R, which corroborates the results of our experiment here.
+Their benchmarks are available in the cli documentation [^4]. So, should we
+be worried about the speed of `cli`? Well, it depends on the context.
+
+Most developers might argue that this is an optimisation overkill[^5]. However,
+we argue that it is important to consider speed differences in context.
+In the case of simple printing, the speed difference is
+negligible yet disruptive and somewhat painful. However, in the grand scheme
+of things, this might be nothing compared to much slower processes that
+need more attention. In those cases, the developer might want to consider other
+optimisation strategies such as profiling [^6]. The essence of this experiment
+is to demonstrate that benchmarking is one way to help developers make design
+decisions.
+
+## Conclusion
+
+In designing R package infrastructure with "heavy" dependencies, it might
+sometimes be necessary to test your intuition about their overhead with
+experiments. Here, we have demonstrated how benchmarking is one way to achieve
+this. We show how a simple decision to use `cli::cli_warn()` to handle
+warnings could come at the cost of a tiny loss in speed, which is worth
+considering in its context.
+
```

Comment:
Would it make sense to link to https://r-pkgs.org/dependencies-mindset-background.html#sec-dependencies-pros-cons for a longer discussion on why every dependency taking decision is unique and how to take context into account?

Reply:
I've linked it now.